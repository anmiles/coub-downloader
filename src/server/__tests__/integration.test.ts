import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { spawn } from 'child_process';

jest.setTimeout(3600000);

async function asyncSpawn(command: Parameters<typeof spawn>[0], args: Parameters<typeof spawn>[1], options: Parameters<typeof spawn>[2]): Promise<number> {
	return new Promise((resolve, reject) => {
		const spawned = spawn(command, args, options);

		spawned.stdout?.on('data', function (data) {
			console.log(data.toString());
		});
		
		spawned.stderr?.on('data', function (data) {
			console.error(data.toString());
		});

		spawned.on('error', reject);
		spawned.on('exit', resolve);
	});
}

function checkFileHash(filename: string) {
	const fileBuffer = fs.readFileSync(filename);
	const hashSum = crypto.createHash('sha256');
	hashSum.update(fileBuffer);

	return {
		filename,
		hash: hashSum.digest('hex')
	};
}

describe('src/server/integration', () => {
	// Launching `npm run build:server` is needed for getting this test passed
	const baseDir = 'src/server/__tests__';
	if (!fs.existsSync(baseDir)) fs.mkdirSync(baseDir);

	const filename = `${baseDir}/coubs.json`;
	const outputDir = `${baseDir}/output`;
	const mediaDir = `${outputDir}/media`;

	it('should work', async () => {
		const cwd = path.resolve(path.join(__dirname, '../../..'));
		const script = 'dist/server/index.js';
		const scriptPath = path.resolve(path.join(cwd, script));

		if (!fs.existsSync(scriptPath)) {
			throw `Compiled script ${scriptPath} does not exist. Execute 'npm run build:server' first`;
		}

		const exitCode = await asyncSpawn('node', [script, filename, outputDir], { cwd });

		expect(exitCode).toBe(0);

		const coubIDs = fs.readdirSync(mediaDir);
		expect(coubIDs).toEqual(['test1', 'test2', 'test3']);

		coubIDs.forEach((coubID) => {
			const coubDir = `${mediaDir}/${coubID}`; 
			fs.readdirSync(coubDir).forEach(file => {
				expect(checkFileHash(`${coubDir}/${file}`)).toMatchSnapshot();
			});
		});

		fs.rmdirSync(outputDir, {recursive: true});
	});
});
