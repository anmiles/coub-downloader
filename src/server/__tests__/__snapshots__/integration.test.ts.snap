// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/server/integration should work 1`] = `
Object {
  "filename": "src/server/__tests__/output/media/test1/test1.jpg",
  "hash": "979cd36f2494ea1cd5fb06859f23f3f3ef017a46a86896b04a884977461c2592",
}
`;

exports[`src/server/integration should work 2`] = `
Object {
  "filename": "src/server/__tests__/output/media/test1/test1.mp4",
  "hash": "96975c27e31beecacbc162898ba8b7a4cd57c1d522cebf2e372ae8e57a71aa61",
}
`;

exports[`src/server/integration should work 3`] = `
Object {
  "filename": "src/server/__tests__/output/media/test2/test2.jpg",
  "hash": "51efbd6c5eef1fd7d25cc74f1704ce3645e1f7c110a0a93a3fa90028acdad7f7",
}
`;

exports[`src/server/integration should work 4`] = `
Object {
  "filename": "src/server/__tests__/output/media/test2/test2.mp3",
  "hash": "2537137ca6972a4bf4f2fcbcca77eb8cea7c72852a19da08b26b5e9737525843",
}
`;

exports[`src/server/integration should work 5`] = `
Object {
  "filename": "src/server/__tests__/output/media/test2/test2.mp4",
  "hash": "a19c9afbaa6cad1af57928176634848f08b9c695aa1c810a011fecab73111a79",
}
`;

exports[`src/server/integration should work 6`] = `
Object {
  "filename": "src/server/__tests__/output/media/test3/test3.jpg",
  "hash": "3f3ab946df1ecadeadcc9d9eaad11ddd4c7efdf9b8f46785ae975e98788169a8",
}
`;

exports[`src/server/integration should work 7`] = `
Object {
  "filename": "src/server/__tests__/output/media/test3/test3.mp3",
  "hash": "69b573ab13dac7314538195912676c1a4900114ed2cba3d5f453b8fd47c2432d",
}
`;

exports[`src/server/integration should work 8`] = `
Object {
  "filename": "src/server/__tests__/output/media/test3/test3.mp4",
  "hash": "3b369d1166b9b5986fb909d1f633180c0d65188321af554cc477d1282bc6524d",
}
`;
